osu.Game/Screens/Multi/RealtimeMultiplayer/RealtimePlayer.cs

```diff
@@ -0,0 +1,81 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System;
+using System.Diagnostics;
+using System.Threading;
+using System.Threading.Tasks;
+using osu.Framework.Allocation;
+using osu.Game.Online.Multiplayer;
+using osu.Game.Online.RealtimeMultiplayer;
+using osu.Game.Scoring;
+using osu.Game.Screens.Multi.Play;
+using osu.Game.Screens.Ranking;
+
+namespace osu.Game.Screens.Multi.RealtimeMultiplayer
+{
+    public class RealtimePlayer : TimeshiftPlayer
+    {
+        protected override bool PauseOnFocusLost => false;
+
+        // Disallow fails in multiplayer for now.
+        protected override bool CheckModsAllowFailure() => false;
+
+        [Resolved]
+        private StatefulMultiplayerClient client { get; set; }
+
+        private readonly TaskCompletionSource<bool> resultsReady = new TaskCompletionSource<bool>();
+        private bool started;
+
+        public RealtimePlayer(PlaylistItem playlistItem)
+            : base(playlistItem, false)
+        {
+        }
+
+        [BackgroundDependencyLoader]
+        private void load()
+        {
+            if (Token == null)
+                return; // Todo: Somehow handle token retrieval failure.
+
+            client.MatchStarted += onMatchStarted;
+            client.ResultsReady += onResultsReady;
+            client.ChangeState(MultiplayerUserState.Loaded);
+
+            while (!started)
+                Thread.Sleep(100);
```

Comment:
slight teeth pain at this but I suppose also maybe not super important for now - can this cause problems? maybe TODO it at least?

Reply:
You've brought up an interesting point though, which is that this should've had a timeout. I've changed it to using an MRE.